### ISSUE TITLE
Debug TypeScript test files - no breakpoints hit (win 10, ts 2.4.2, node 8.2.1, vs code latest)

### ISSUE THREAD

**Issue by jstott** [ospatil/generator-node-typescript#35]
First off - awesome generator, thanks for all of your efforts.
Everything works perfectly, except the ability to debug the test file(s).

Here is what I saw when I launch test 'Debug test' with the greeter-spec.ts file open in the editor:

```Cannot launch program 'd:\VSTS-DevOps\genSample\node_modules\.bin\jest'; setting the 'outDir' or 'outFiles' attribute might help.```

So after several iterations & looking around at similar issues, this is the launch.json section that's working for me: 

```   
 {
      "type": "node",
      "request": "launch",
      "name": "Debug test-v4 works-single-selected file",
      "runtimeArgs": [
        "--inspect-brk",
        "${workspaceRoot}/node_modules/jest/bin/jest.js",
        "--findRelatedTests",
        "${relativeFile}",
        "--env",
        "jest-environment-node-debug",
        "--runInBand"
      ],
      "console": "integratedTerminal"
    }
```

I tried various combinations of program = .../.bin/jest.cmd, .../jest/bin/jest.js yada yada- but did not hit on anything until the runtimeArgs vs program & args options setup.

Interested in your take, or if anyone else ran into something similar...

working env: Windows 10, TS 2.4.2, node 8.2.1, vs code latest.

**Comment by ospatil** [comment]
When it comes to windows, I'm afraid it's really a shot in the dark for me. I don't have means to test it on Windows and rely on feedback and pull requests from good users like yourself instead :).

The `launch.json` configuration you shared above, is it fully working configuration? In such case, I can update the generator to create this when the OS is windows?

**Comment by jstott** [comment]
Thanks for the feedback.  The config above is a fully working config.  That said, if no one else has reported an issue for windows or validated the proposed fix besides me, let me engage a few folks for testing in various combo's to ensure this isn't a 'works on my machine' windows type thing.  Will update on those results.

**Comment by GaryB432** [comment]
This fixed the same issue for me too. Looking forward to your "engaging a few folks" for more testing. A shame we can't have a cross-plat task and @ospatil anticipates having to generate a different task (yours as above?) for windows. That will obviously make the generated app platform-specific. But I guess that can't be avoided.

**Comment by GaryB432** [comment]
The above task worked for me too.  Looking forward to the results of your "engaging".  Thanks for having windows friends to try "various combos"!

**Comment by jstott** [comment]
Below is the most recent iteration - modified from what I posted previously (that flavor had issues with   envFile file options).  Unfortunatly, there are issues depending on the version of node and whether or not jest-cli is installed or not.  I need to work up a matrix for the node versions.

With TS 2.4.2, node 8.2.1, VS Code latest (as noted above, not working for node 6.10 / 6.11)

```
{
      "type": "node",
      "request": "launch",
      "name": "Debug test",
      "program": "${workspaceRoot}/node_modules/jest-cli/bin/jest.js",
      "args": [
        "--findRelatedTests",
        "${relativeFile}",
        "--env",
        "jest-environment-node-debug"
      ],
      "cwd": "${workspaceRoot}"
    }
```
Will update with a better list & details as I compile it.

**Comment by GaryB432** [comment]
Nice work! Glad I'm an early adopter so I can use the early finding

**Comment by ospatil** [comment]
@jstott nice work! The configuration works on mac too, so no need to have OS-specific separate configs. Stay tuned, this will be part of 4.1.0 release on this weekend.

**Comment by davidsk** [comment]
@ospatil FYI I had to install jest-cli and upgrade to latest node before I could get breakpoints in typescript to work.  Environment: Windows 10, TS 2.5.3, node 8.6.0, vscode 1.17.0

Can anyone explain why the jest-cli isn't necessary?  It's explicitly referenced in the launch.json

**Comment by ospatil** [comment]
@davidsk I'm sorry for the delayed response. regarding `jest-cli`, I suppose it's a `yarn` vs `npm` thing because `yarn` installs `jest-cli` and it's available for use.

I think `jest-cli` may not be needed at all. I pointed the test launch program to `jest.js` under `jest` as shown below and it worked just fine:
```js
"name": "Debug test",
"program": "${workspaceRoot}/node_modules/jest/bin/jest.js",
```

Could you please try it out and let me know?

**Comment by donpedro** [comment]
I am running under Win10, installing with npm and now under the latest launch.js configuration as installed (and listed by @jstott above). My experience:

- jest-cli is missing; switching to use jest.js under jest as indicated by @ospatil works. (possible fix: refer to jest out of the box instead of jest-cli to bypass this problem; see #45)
- even after above change, no breakpoints are hit when using Node v6.1. Solution appears related to node.js debugging changes: apparently this type of debugging depends on the "Inspector" protocol. According to [this](https://code.visualstudio.com/docs/nodejs/nodejs-debugging), the Inspector protocol is available in Windows in Node v6.9, however in my tests 7.0 didn't work, but 8.0 did. 8.0 is the first version that defaults to use of the Inspector protocol.

I went to Node v6.9 and added the following line to the launch configuration object: 

`"protocol": "inspector"`

This forces use of Inspector protocol, and breakpoints now work.
